proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+- Flexible enough abstractions to explore the problem space with multiple solutions
+- Appropriate licensing
+
+`servant` (and the `wai` ecosystem in general) is a success story and an example of higher-order programming making something safer and easier, something that we can aspire to do with cryptography.
+
+## Technical Content
+
+This proposal is for the contruction of a suite of cryptography libraries, ranging from low-level bindings to specific algorithms, to higher-level abstractions and typeclasses.
+
+This leg of the proposal focuses on bindings to the `botan` cryptography, in the form of the following libraries:
+
+### botan-bindings
+
+The `botan-bindings` library will contain raw bindings and is an almost direct, 1-1 translation of the C API into Haskell FFI calls. As such, it exposes and operates over C FFI types, requires buffer and pointer and pointer management, and returns error codes. This library only exposes FFI calls and constants, and is suitable for building your own abstraction over Botan.
+
+### botan-low
+
+The `botan-low` library will contain low-level bindings which wrap the FFI calls into monadic IO actions. This library handles the translation between buffers and ByteStrings, and throws exceptions in the case of errors, but is otherwise a fairly faithful translation of the interface. Thus this library is suitable for use, but is not recommended for day-to-day programming due to statefulness and a lack of type safety.
+
+### botan
+
+The `botan` library will contain high-level bindings which hides the stateful IO and exposes appropriately idiomatic and pure interfaces. It will provide data types for algorithms, enumerations for constants, convenience methods for key and nonce generation, incremental processing of lazy bytestrings, and otherwise bring significant safety and ergonomics over `botan-low` and `botan-bindings`.
+
+## Optional / Future Technical Content
+
+More advanced concepts such as typeclasses may be delivered in a future leg focused on generalization, but some development occur earlier in order to provide a testable 'use case' for the stack of libraries. If the first leg is completed ahead of schedule, the remaining effort will go towards these libraries.
+
+### crypto-schemes
+
+The `crypto-schemes` library will contain a hierarchy of abstract type classes and data families for cryptography and scheme identifiers, allowing for a unified interface to multiple back-ends. It will also provide various utility functions including sized bit and byte vectors, block and padding functions.
+
+### crypto-schemes-botan
+
+The `crypto-schemes-botan` library will be an implementation of `crypto-schemes` using `botan`, implementing class instances for the specific algorithms that `botan` provides.
+
+### botanium
+
+The `botanium` library will contain user-friendly bindings to specific algorithms, designed for ease of use much like `libsodium`, but using `botan` or `crypto-schemes-botan` as a backend.
```

Comment:
I'm afraid that this name will be understood as a reference to "Chromium" rather than "Sodium", as the free version of an otherwise-proprietary program.

Reply:
The higher-level abstractions / libraries are being struck from this proposal to increase clarity / narrow scope / avoid confusion as they aren't the focus of this proposal.